jetty-util-ajax/src/main/java/org/eclipse/jetty/util/ajax/AsyncJSON.java

```diff
@@ -0,0 +1,1205 @@
+        // We are done with the field.
+        stack.pop();
+        @SuppressWarnings("unchecked")
+        Map<String, Object> map = (Map<String, Object>)stack.peek().value;
+        map.put(name, value);
+
+        return true;
+    }
+
+    private Object convertObject(Map<String, Object> object)
+    {
+        Object result = convertObject("x-class", object);
+        if (result == null)
+        {
+            result = convertObject("class", object);
+            if (result == null)
+                return object;
+        }
+        return result;
+    }
+
+    private Object convertObject(String fieldName, Map<String, Object> object)
+    {
+        String className = (String)object.get(fieldName);
+        if (className == null)
+            return null;
+
+        JSON.Convertible convertible = toConvertible(className);
+        if (convertible != null)
+        {
+            convertible.fromJSON(object);
+            return convertible;
+        }
+
+        JSON.Convertor convertor = factory.convertors.get(className);
+        if (convertor != null)
+            return convertor.fromJSON(object);
+
+        return null;
+    }
+
+    private JSON.Convertible toConvertible(String className)
+    {
+        try
+        {
+            Class<?> klass = Loader.loadClass(className);
+            if (JSON.Convertible.class.isAssignableFrom(klass))
+                return (JSON.Convertible)klass.getConstructor().newInstance();
+            return null;
+        }
+        catch (Throwable x)
+        {
+            throw new IllegalArgumentException(x);
+        }
+    }
+
+    protected RuntimeException newInvalidJSON(ByteBuffer buffer, String message)
+    {
+        Utf8StringBuilder builder = new Utf8StringBuilder();
+        builder.append(System.lineSeparator());
+        int position = buffer.position();
+        if (factory.isDetailedParseException())
+        {
+            chunks.forEach(chunk -> builder.append(buffer));
+        }
+        else
+        {
+            buffer.position(0);
+            builder.append(buffer);
+            buffer.position(position);
+        }
+        builder.append(System.lineSeparator());
+        String indent = "";
+        if (position > 1)
+        {
+            char[] chars = new char[position - 1];
+            Arrays.fill(chars, ' ');
+            indent = new String(chars);
+        }
+        builder.append(indent);
+        builder.append("^ ");
+        builder.append(message);
+        return new IllegalArgumentException(builder.toString());
+    }
+
+    public interface Context
+    {
+        public int depth();
+    }
+
+    private enum State
+    {
+        COMPLETE, NULL, TRUE, FALSE, NUMBER, STRING, ESCAPE, UNICODE, ARRAY, OBJECT, OBJECT_FIELD, OBJECT_FIELD_NAME, OBJECT_FIELD_VALUE
+    }
+
+    private static class Frame
+    {
+        private String name;
+        private State state;
+        private Object value;
+
+        private Frame(State state, Object value)
```

Comment:
This constructor is useless as it's only ever called with `(null, null)`

Reply:
Removed.